server/src-lib/Hasura/GraphQL/Schema/Remote.hs

```diff
@@ -0,0 +1,424 @@
+       , Maybe [P.FieldParser n (RemoteSchemaInfo, Field NoFragments Variable)])
+buildRemoteParser (IntrospectionResult sdoc query_root mutation_root subscription_root) info = do
+  queryT <- makeParsers query_root
+  mutationT <- traverse makeParsers mutation_root
+  subscriptionT <- traverse makeParsers subscription_root
+  return (queryT, mutationT, subscriptionT)
+  where
+    makeFieldParser :: G.FieldDefinition -> m (P.FieldParser n (RemoteSchemaInfo, Field NoFragments Variable))
+    makeFieldParser fieldDef = do
+      fp <- remoteField' sdoc fieldDef
+      return $ do
+        raw <- P.unsafeRawField (P.fDefinition fp)
+        return (info, raw)
+    makeParsers :: G.Name -> m [P.FieldParser n (RemoteSchemaInfo, Field NoFragments Variable)]
+    makeParsers rootName =
+      case lookupType sdoc rootName of
+        Just (G.TypeDefinitionObject o) ->
+          traverse makeFieldParser $ _otdFieldsDefinition o
+        _ -> throw400 Unexpected $ rootName <<> " has to be an object type"
+
+-- | 'remoteFieldFullSchema' takes the 'SchemaIntrospection' and a 'G.Name' and will
+--   return a 'SelectionSet' parser if the 'G.Name' is found and is a 'TypeDefinitionObject',
+--   otherwise, an error will be thrown.
+remoteFieldFullSchema
+  :: forall n m
+   . (MonadSchema n m, MonadError QErr m)
+  => SchemaIntrospection
+  -> G.Name
+  -> m (Parser 'Output n (G.SelectionSet NoFragments Variable))
+remoteFieldFullSchema sdoc name =
+  P.memoizeOn 'remoteFieldFullSchema name do
+  fieldObjectType <-
+    case lookupType sdoc name of
+      Just (G.TypeDefinitionObject o) -> pure o
+      _ -> throw400 RemoteSchemaError $ "object with " <> G.unName name <> " not found"
+  fieldParser <- remoteSchemaObject sdoc fieldObjectType
+  pure $ P.unsafeRawParser (P.pType fieldParser)
+
+remoteField'
+  :: forall n m
+  . (MonadSchema n m, MonadError QErr m)
+  => SchemaIntrospection
+  -> G.FieldDefinition
+  -> m (FieldParser n ())
+remoteField' schemaDoc (G.FieldDefinition description name argsDefinition gType _) =
+  let
+    addNullableList :: FieldParser n () -> FieldParser n ()
+    addNullableList (P.FieldParser (Definition name' un desc (FieldInfo args typ)) parser)
+      = P.FieldParser (Definition name' un desc (FieldInfo args (Nullable (TList typ)))) parser
+
+    addNonNullableList :: FieldParser n () -> FieldParser n ()
+    addNonNullableList (P.FieldParser (Definition name' un desc (FieldInfo args typ)) parser)
+      = P.FieldParser (Definition name' un desc (FieldInfo args (NonNullable (TList typ)))) parser
+
+    -- TODO add directives, deprecation
+    convertType :: G.GType -> m (FieldParser n ())
+    convertType gType' = do
+        case gType' of
+          G.TypeNamed (Nullability True) fieldTypeName -> do
+            remoteFld <- remoteFieldFromName schemaDoc name description fieldTypeName argsDefinition
+            pure . P.nullableField $ remoteFld
```

Comment:
nit: this can be rewritten as `P.nullableField <$> remoteFieldFromName ...`, and without the `do`, like in the `TypeList` case below.

Reply:
Yes, I'll make this change

Reply:
Done